src/ray/common/memory_monitor.cc

```diff
@@ -0,0 +1,185 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/memory_monitor.h"
+
+#include <fstream>  // std::ifstream
+#include <tuple>
+
+#include "ray/common/nullable_uint.h"
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+
+MemoryMonitor::MemoryMonitor(float usage_threshold_,
+                             uint64_t monitor_interval_ms,
+                             MemoryUsageRefreshCallback monitor_callback)
+    : usage_threshold_(usage_threshold_),
+      monitor_callback_(monitor_callback),
+      io_context_(),
+      monitor_thread_([this] {
+        boost::asio::io_service::work io_service_work_(io_context_);
+        io_context_.run();
+      }),
+      runner_(io_context_) {
+  RAY_CHECK(monitor_callback_ != nullptr);
+  RAY_CHECK_GE(usage_threshold_, 0);
+  RAY_CHECK_LE(usage_threshold_, 1);
+  if (monitor_interval_ms > 0) {
+#ifdef __linux__
+    runner_.RunFnPeriodically(
+        [this] {
+          bool is_usage_above_threshold = IsUsageAboveThreshold();
+          monitor_callback_(is_usage_above_threshold);
+        },
+        monitor_interval_ms,
+        "MemoryMonitor.CheckIsMemoryUsageAboveThreshold");
+    RAY_LOG(INFO) << "MemoryMonitor initialized";
+#else
+    RAY_LOG(WARNING) << "Not running MemoryMonitor. It is currently supported "
+                     << "only on Linux.";
+#endif
+  } else {
+    RAY_LOG(INFO) << "MemoryMonitor disabled. Specify "
+                  << "`memory_monitor_interval_ms` > 0 to enable the monitor.";
+  }
+}
+
+bool MemoryMonitor::IsUsageAboveThreshold() {
+  auto [used_memory_bytes, total_memory_bytes] = GetMemoryBytes();
+  if (total_memory_bytes == kNull || used_memory_bytes == kNull) {
+    RAY_LOG(WARNING) << "Unable to capture node memory. Monitor will not be able "
```

Comment:
```suggestion
    RAY_LOG_EVERY_N(WARNING, 100) << "Unable to capture node memory. Monitor will not be able "
```
Could be too spammy when something actually goes wrong?

Reply:
sg - since this runs periodically I am using RAY_LOG_EVERY_MS